### PR TITLE
Initial integration of the provider dependency pinning work

### PR DESCRIPTION
**Pull request by apparentlymart**
This follows on from some earlier work that introduced models for representing provider dependency "locks" and a file format for saving them to disk. This PR wires the new models and behaviors into the `terraform init` command, making it use the new user-facing lock file `.terraform.lock.hcl` in place of the old implementation-detail `.terraform/plugin/selections.json` file.

This new file is intended to be checked in to version control. Although it's primarily intended for automatic updates rather than manual updates, we've switched to using HCL syntax for it because humans will see changes to the lock file as part of pull requests and (subjectively) diffs of the HCL-based locks language seem easier to read than diffs of a JSON data structure.

I still have work to do here in testing more edge cases, polishing the user experience, and adding test coverage for some behaviors that apparently weren't previously tested. However, I've been on a merge conflict treadmill for the past week due to working on this concurrently with code cleanup and refactoring, so I'd like to merge this checkpoint and then continue working on the details in subsequent PRs.

Because of that, I'd primarily like to focus this review on whether the functionality here seems to do what it sets out to do, and de-emphasize broader feedback about exactly how this fits into the workflow, how the UI behaves, etc. Both forms of feedback are welcome, but I'd ask you to be clear about which category your feedback fits into so I can understand what is blocking merging _this_ PR and what I can address in future PRs. Ultimately my goal is to merge this soon before it becomes merge-conflicted again, addressing as much feedback as possible in separate PRs.

I specifically intend to make follow-up PRs with additional text coverage for the new command added here and for various `terraform init` provider installation situations that currently lack test coverage, but I decided to separate that just to avoid the drag of continuing merge conflicts.

I've included in this PR [a first draft of the user-facing documentation for this](https://github.com/hashicorp/terraform/blob/f-provider-dep-pinning/website/docs/configuration/dependency-lock.html.md) primarily because I think that's the best way to show the overall behavior I intend this to have. I expect that documentation to see at least one more editing revision after this though, so that I can do it with my "technical writing brain" inserted, rather than with my "programming brain".

### CODE REVIEW

**Review by mildwonkey — APPROVED** [review]
This is great and I'm really excited to get it out in the world. 

All of my inline comments are very small thoughts, not related to the actual functionality, and nothing that should stop you from merging this.

I have one bit of workflowicious feedback (which still doesn't stop this from being merged). Since this is a new feature that (slightly) changes the `terraform init` behavior, could we consider an "information only" mode for the first release (either by default, or via a command-line option, printing warnings during `init` instead of errors? I'm certainly ok with not doing this, or at least waiting to see what kind of feedback we actually get, but it also might help save us some user frustration if they run into weird edge cases (or bugs, now that I think of it).

File: website/docs/commands/providers/lock.html.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+in order to get access to the most detailed information about the provider
+packages.
+
+To create a lock entry for a particular provider that is available only in a
+local mirror, you can use either the `-fs-mirror` or `-net-mirror` command
+line options to override the default behavior of consulting the provider's
+origin registry:
+
+```
+terraform providers lock \
```
[the remark this answers is not in the document]
Comment (apparentlymart) [reply website/docs/commands/providers/lock.html.md:132]: I had originally conceived this command as a way to temporarily _ignore_ the installation methods in the CLI config in order to generate checksums in spite of the mirrors, and so I had originally thought that such an option would be superfluous because it would just mimic what `terraform init` does by default, but on reflection I suppose the combination of that option and multiple `-platform` could be interesting for the case of "I want to pre-lock multiple platforms", if you are using an installation method configuration that only overrides one or two in-house providers.

In other words, I agree with you 😀 but I will hold on including that in this PR because I want to think a bit more about what to call this option.

File: internal/depsfile/testing.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	"github.com/google/go-cmp/cmp"
+)
+
+// ProviderLockComparer is an option for github.com/google/go-cmp/cmp that
+// specifies how to compare values of type depsfile.ProviderLock.
+//
+// Use this, rather than crafting comparison options yourself, in case the
+// comparison strategy needs to change in future due to implementation details
+// of the ProviderLock type.
+var ProviderLockComparer cmp.Option
```
[the remark this answers is not in the document]
Comment (apparentlymart) [reply internal/depsfile/testing.go:13]: I think I got this from reading some issues in the `go-cmp` repository. Seemed like a nice pattern to keep callers from depending on the implementation details. 😀

File: command/meta_providers.go [hunk 3]
```diff
@@ -160,34 +161,86 @@ func (m *Meta) providerInstallSource() getproviders.Source {
...
+	cacheDir := m.providerLocalCacheDir()
+
 	// The internal providers are _always_ available, even if the configuration
 	// doesn't request them, because they don't need any special installation
 	// and they'll just be ignored if not used.
 	internalFactories := m.internalProviders()
 
-	factories := make(map[addrs.Provider]providers.Factory, len(selected)+len(internalFactories)+len(m.UnmanagedProviders))
+	// The Terraform SDK test harness (and possibly other callers in future)
+	// can for use of their own already-started provider servers, which we
```
[the remark this answers is not in the document]
Comment (apparentlymart) [reply command/meta_providers.go]: Thanks!
